presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ImplementIntersectAsUnion.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.ExpressionUtils;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.IntersectNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.tree.ComparisonExpression;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.GenericLiteral;
+
+import java.util.List;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.prestosql.sql.planner.plan.Patterns.intersect;
+import static io.prestosql.sql.tree.ComparisonExpression.Operator.GREATER_THAN_OR_EQUAL;
+
+/**
+ * Converts INTERSECT queries into UNION ALL..GROUP BY...WHERE
+ * E.g.:
+ * <pre>
+ *     SELECT a FROM foo
+ *     INTERSECT
+ *     SELECT x FROM bar
+ * </pre>
+ * =>
+ * <pre>
+ *     SELECT a
+ *     FROM
+ *     (
+ *         SELECT a,
+ *         COUNT(foo_marker) AS foo_count,
+ *         COUNT(bar_marker) AS bar_count
+ *         FROM
+ *         (
+ *             SELECT a, true as foo_marker, null as bar_marker
+ *             FROM foo
+ *             UNION ALL
+ *             SELECT x, null as foo_marker, true as bar_marker
+ *             FROM bar
+ *         ) T1
+ *     GROUP BY a
+ *     ) T2
+ *     WHERE foo_count >= 1 AND bar_count >= 1;
+ * </pre>
+ */
+public class ImplementIntersectAsUnion
+        implements Rule<IntersectNode>
+{
+    private static final Pattern<IntersectNode> PATTERN = intersect();
+
+    @Override
+    public Pattern<IntersectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(IntersectNode node, Captures captures, Context context)
+    {
+        SetOperationNodeTranslator translator = new SetOperationNodeTranslator(context.getSymbolAllocator(), context.getIdAllocator());
+        SetOperationNodeTranslator.TranslatorResult result = translator.simplify(node);
+
+        List<Expression> predicates = result.getAggregationNode().getAggregations().keySet().stream()
```

Comment:
Another idea that should help clean up the abstraction further. Instead of returning counts, make it return booleans that indicate whether the row is "present" in each of the branches of the set operation. This class would just add a filter that says `present_1 AND present_2 AND present_3 AND ...`. `EXCEPT` would, on the other hand, do `present_1 AND NOT present_2`.